fud2/fud-core/src/run.rs

```diff
@@ -147,13 +162,66 @@ impl EmitBuild for OpEmitData {
             }
         }
 
-        emitter.build_cmd_with_vars(
-            outputs,
-            &self.op_name,
-            inputs,
-            &[],
-            &vars,
-        )?;
+        // Some outputs are created by user specified `shell_deps`. Others are not and have to be
+        // created by running every `shell`/`shell_deps` command.
+        let mut uncreated_outputs: HashSet<String> =
+            HashSet::from_iter((0..outputs.len()).map(|i| {
+                format!("${}", crate::run::io_file_var_name(i, false))
+            }));
+
+        // Write a sequence of rules for each cmd.
+        for (i, cmd) in self.cmds.iter().enumerate() {
+            let rule_name = format!("{}_rule_{}", self.op_name, i + 1);
+            emitter.rule(&rule_name, &cmd.cmd)?;
+            let targets = cmd.gens.iter().map(|s| s.as_str()).collect_vec();
+            let deps = cmd.deps.iter().map(|s| s.as_str()).collect_vec();
+            for &t in &targets {
+                uncreated_outputs.remove(t);
+            }
+            emitter.build_cmd_with_vars(
+                &targets,
+                &rule_name,
+                &deps,
+                &[],
+                &vars,
+            )?;
+        }
+
+        // Depend on all shell commands. The illusion is all shell commands are run in order, with
+        // the option to specify some shell commands as having explicit targets and dependancies as
+        // an optimization. They still all have to get run though.
+        let deps = inputs
+            .iter()
+            .cloned()
+            .chain(
+                self.cmds
+                    .iter()
+                    .map(|cmd| cmd.gens.last().unwrap().as_str()),
+            )
+            .collect_vec();
+
+        if uncreated_outputs.is_empty() {
+            // Even if all outputs have said to have been created by the user, all the calls to `shell`
+            // should still get run.
+            let fake_output = format!("_{}_output.fake", self.op_name);
+            emitter.build_cmd_with_vars(
+                &[&fake_output],
+                "phony",
+                &deps,
+                &[],
+                &vars,
+            )?;
+            writeln!(emitter.out, "default {}", fake_output)?;
```

Comment:
I'm reasonably sure we do not want to use a `default` to make this work. This will make the op run unconditionally in incremental builds, even if we only need unrelated files.

I see the problem you're grappling with here. What if someone mixes `shell` commands and `shell_deps` commands, and they are both careful (in the sense that they explicitly declare what produces all outputs) and sloppy (they throw some `shell` commands in there hoping that they also get run, despite the lack of dependency information). You want a way to make those "plain" `shell` command run at _some point_, but there are no remaining output files available to foist arbitrary dependencies upon to make that happen. All those output files are already being produced by other commands!

Broadly, this is making me feel bolder about the assertion that it is not very useful or important to allow mixing of `shell` and `shell_deps`… for a given op, it would simplify things a lot of people had to choose between being careful and sloppy and could not mix the two. How heartbroken would you be if we just disallowed this entirely? I really don't think people will want to mix the two styles like this.

Reply:
Yeah...
The more I'm thinking about this, I'm starting to agree. 

The reason I'm resistant is that this makes effectively two types of ops, one where shell commands have dependencies and one where they don't, one type sugar for the other. I guess the tradeoff there is the what should happen in each case becomes super clear. That probably is better than the somewhat ambiguous mess of combining everything.

Yeah, the example which I think you are also seeing of having `shell_deps` commands generate all required outputs but still having a `shell` command is what makes clear these things should be separate. Otherwise it is just arbitrary and non-obvious what to run.

Maybe another way of looking at it is the user has to choose between the functional style of `shell_deps` or the more imperative style of `shell` when writing ops. To mix the two requires some concrete interface, a purpose of ops.

Reply:
> The reason I'm resistant is that this makes effectively two types of ops, one where shell commands have dependencies and one where they don't, one type sugar for the other.

Yeah, you're absolutely right about this. We could even decide to implement them separately… as in, there could be a `SeqRulesOp` and a general `RulesOp`. Not sure if that's better or worse than doing the more sugar-like thing that you suggest.

Reply:
In my mind, `RulesOp` models Ninja, not Rhai, so if there is going to be two different structs it would be varying `RhaiOp` (in `plugin.rs`), the struct which gets turned into a `RulesOp` (especially because `SeqRulesOp` would just be a special case of `RulesOp`.

The second thing might make sense, but I'm just going to go with one struct for both for now.

Reply:
That sounds right to me.